Add a "requires" directive, extend functionality of package requirements

alalazo opened the pull request:
closes #17344
closes #21765

This PR adds a new directive:
```python
requires("%gcc", when=..., msg=...)
```
which behaves the opposite of `conflicts`. The directive uses the same mechanism as requirements in `packages.yaml`, so these have been extended to support a `when` and a `message` attribute too in the configuration file. 

Modifications:
- [x] Add a `requires` directive
- [x] Allow for conditional requirements using a `when=` argument
- [x] Allow for custom error messages
- [x] Add unit tests and docs

alalazo committed:
Add a "requires" directive

The "requires" directive at the moment takes a
single string, and is equivalent to the string
form of the same requirement in packages.yaml

alalazo committed:
Collect attributes of a requirement rule together

alalazo committed:
Use an enum for provenance in facts

alalazo committed:
The requires directive support a condition and a custom message

This puts the directive on par with 'conflicts', since it is
its logical opposite.

Hook custom messages into constraint resolution.

Add the same arguments also to the extended syntax in packages.yaml

alalazo committed:
Add logic to allow conditional requirements

alalazo committed:
Constrain the platforms that can be used in a solve

alalazo committed:
Add a few unit tests, express rules in a way that fits our error handling

alalazo committed:
Added docs

alalazo committed:
Fix issues with error priorities

With the new setup, we should see custom error messages
to avoid high priority internal error messages talking
about "two compilers" being chosen for the spec.

alalazo commented:
I checked with the usual radiuss benchmark:

[radiuss_requires.csv](https://github.com/spack/spack/files/11040287/radiuss_requires.csv)
[radiuss_develop.csv](https://github.com/spack/spack/files/11040288/radiuss_develop.csv)
[radiuss.txt](https://github.com/spack/spack/files/11040289/radiuss.txt)

After a couple of small changes, it seems that the new encoding is _slightly_ faster too (in absence of requirements). Plots are on `develop`:

![develop](https://user-images.githubusercontent.com/4199709/226916468-0534e091-06b6-47a0-ab88-8260d37778ab.png)

and on this PR:

![requires](https://user-images.githubusercontent.com/4199709/226916536-2e944a28-efcd-4fa5-95bf-9b8e5674a2db.png)

mwkrentel commented:
@alalazo  @hainest 
You may want to consider adding the `one_of` and `any_of` features at
some point.  For example, both dyninst and hpctoolkit are working to
being compiled with either gcc or llvm.  We're not there yet, but
we're working on it.

Also, intel-xed requires 64- or 32-bit x86 (arch x86_64 or x86).
(Well, no one uses 32-bit and it probably doesn't build there,
but in principle it could.)

alalazo committed:
Resolved conflicts with #36347

alalazo commented:
@scheibelp @tgamblin Following our discussion, I'm posting here an example of a conditional requirement that _might_ behave in a surprising way, even though it works as expected. Let's start by establishing what `zlib` concretizes to in my environment without any requirements:
```console
$ spack spec zlib
Input spec
--------------------------------
zlib

Concretized
--------------------------------
zlib@1.2.13%gcc@9.4.0+optimize+pic+shared build_system=makefile arch=linux-ubuntu20.04-icelake
```
so it uses `@1.2.13` and `target=icelake`. Node attributes in the spec are subject to a multi-objective optimization with strict priorities among the parts:

![2023-03-30_10-28](https://user-images.githubusercontent.com/4199709/228776816-fa815489-75e2-45c2-9357-09840ae365b7.png)

If we:
1. Write a requirement for a higher level attribute, requiring a non-default value (so we get a penalty in the optimization)
2. Make the requirement conditional to a lower-level default value 

`clingo` will swap the value to the next best for the conditional, to avoid taking a penalty on a lower level criteria. Practical example:
```yaml
packages:
  zlib:
    require:
    - one_of: ["@1.2.12"]
      when: "target=icelake"
```
Here I require an old version of `zlib` if the target is `icelake`:
```console
$ spack spec zlib
Input spec
--------------------------------
zlib

Concretized
--------------------------------
zlib@1.2.13%gcc@9.4.0+optimize+pic+shared build_system=makefile arch=linux-ubuntu20.04-cascadelake
```
When I concretize `clingo` swaps `icelake` for `cascadelake`, so to take a penalty there (lowest priority in the optimization criteria) instead of taking it on the version. 

**TL;DR** The mechanism is general, and I previously made an example with a compiler instead of a version. We can construct multiple examples leading to the same behavior if we use a lower level optimization target as the conditional constraint. If this behavior seems fine, and maybe only worth being documented, then the PR is ready.

scheibelp commented:
> If this behavior seems fine, and maybe only worth being documented, then the PR is ready.

This looks fine (more than that, to me it looks like the best choice in this case).

tgamblin commented:
IMO this is expected, *given the precedence of optimization criteria*.

One observation: if you took the same requirement and turned it around, you *wouldn't* get weirdness.  i.e., if you required `target=foobar` when `@1.2.13`, it *wouldn't* end up avoiding the effect of the requirement, because the effect is lower precedence than the trigger.

IMO there are two ways to deal with this:
1. Document the precedence of our optimization criteria and mention this.  Basically say that if your `when:` condition is lower precedence than the requirement, the effect is going to be that the optimizer avoids the requirement unless it exhausts all "better" options than letting the condition be true.  Emphasize that conditional requirements are *not* simple "requests" and they may affect the solver's choice of solution.
2. Make requirements zero-cost so that they "augment" solutions rather than changing their cost vector.  I think we agree that this is more complicated and we can't do it immediately, and I think it'll *still* be confusing to people because your requirement *may* still alter the solver's choice.  If the requirement makes an entire set of solutions unsatisfiable (instead of just "worse" as above) then the solver's going to pick a satisfiable solution instead.

So I think going with (1) is best -- for now, and we should document the precedence.  If people still find this confusing we can think about how to implement (2) but I think it's still going to be very hard to make it predictable.

alalazo commented:
> One observation: if you took the same requirement and turned it around, you wouldn't get weirdness. i.e., if you required target=foobar when @1.2.13, it wouldn't end up avoiding the effect of the requirement, because the effect is lower precedence than the trigger

Yes, that's exactly why I thought it was worth pointing out this behavior explicitly. I think a small fraction of even power users are aware of the optimization criteria. I'll go with 1 and ping for a review, thanks!

alalazo committed:
Add a note on optimization criteria

scheibelp commented on lib/spack/spack/test/concretize_requirements.py:
```diff
@@ -586,3 +587,101 @@ def test_non_existing_variants_under_all(concretize_scope, mock_packages):
+            [("%clang", False), ("%gcc", True)],
+        ),
+    ],
+)
```
Not specifically an issue with (only) these changes, but I find these data structures embedded into parameterization lists to be hard to read. Maybe something like:

```
packages_yaml1 = """
"""
spec_str1 = ""
expected_satisfies1 = False
packages_yaml2 = """..."""

parameters = [(packages_yaml1, spec_str1, expected_satisfies1), ...]
```

scheibelp commented on lib/spack/spack/test/concretize_requirements.py:
```diff
@@ -586,3 +587,101 @@ def test_non_existing_variants_under_all(concretize_scope, mock_packages):
+        ),
+    ],
+)
+def test_requirements_from_packages_yaml(
```
This test name should reflect the fact that we are testing `when` (since other tests already test basic requirements from packages.yaml).

alalazo replied:
Done in 8c85aa790ee87ee77173ab6e98df5dc5019261c6

scheibelp commented on lib/spack/spack/test/concretize_requirements.py:
```diff
@@ -586,3 +587,101 @@ def test_non_existing_variants_under_all(concretize_scope, mock_packages):
+            """\
+    packages:
+      all:
+        compiler: ["gcc", "clang"]
```
(question) is this essential to the test?

alalazo replied:
Yes, in the sense that if the requirement is disregarded we'll be preferring `gcc` to `clang`. Otherwise there is the possibility of selecting `clang` accidentally.

scheibelp replied:
Thanks: a comment explaining that would be useful.

alalazo replied:
Done in 57cd6ed

scheibelp commented on lib/spack/spack/test/concretize_requirements.py:
```diff
@@ -586,3 +587,101 @@ def test_non_existing_variants_under_all(concretize_scope, mock_packages):
+    packages:
+      libelf:
+        require:
+        - one_of: ["%clang"]
```
Can we modify the schema to allow for another key like `spec: "%clang"` if there is only one spec?

alalazo replied:
I don't mind doing that, but can we do it in another PR ? I think people also asked for some syntactic sugar for:
```yaml
- any_of: [..., "@:"]
```
which acts as a much stronger spec preference. So I am wondering if we could combine the two in another PR focused on the YAML configuration.

scheibelp replied:
> I don't mind doing that, but can we do it in another PR ?

Yes

alalazo replied:
In the end I did it in 9a495af5195824b9dda1e81120f29a4657796a68

scheibelp commented on lib/spack/spack/test/concretize_requirements.py:
```diff
@@ -586,3 +587,101 @@ def test_non_existing_variants_under_all(concretize_scope, mock_packages):
+        packages:
+          libelf:
+            require:
+            - one_of: ["%clang"]
```
Might be worth adding a test where we require `one_of: [%gcc, %clang]` and then concretize with `%intel`).

alalazo replied:
Done in 39822a46e4652e1ce297578a54d10d84a90f7966

scheibelp commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1700,6 +1770,7 @@ def platform_defaults(self):
         self.gen.h2("Default platform")
         platform = spack.platforms.host()
         self.gen.fact(fn.node_platform_default(platform))
+        self.gen.fact(fn.allowed_platform(platform))
```
I'm guessing this is required or the `test_requirements_fail_with_custom_message` fails for the test with the `platform=test` condition? If I understand, this means you cannot ever specify another platform (we only generate this fact for the host platform, and `
 :- attr("node_platform", _, Platform), not allowed_platform(Platform).` would cause a failure for any other.

alalazo replied:
> I'm guessing this is required or the test_requirements_fail_with_custom_message fails for the test with the platform=test condition?

I'll double check the details, but this was required because we don't treat `platform` like the other node attributes (which have a list of allowed values they can take). So, if we have requirements based on `platform` that cannot be satisfied, sometimes Spack would pick a wrong platform depending on unrelated facts in the solve.

scheibelp replied:
My assessment is that this is a new hard constraint but that it won't impact users: there are not many platforms and users aren't currently cross compiling other than possibly OS/OS or target/target (in particular since Cray/Linux are now the same platform). My original concern was that a user might be specifying an alternative platform and that this would break that, but I don't think they would get anything out of doing so, and could therefore reconcretize with their local platform without issue.

scheibelp commented on lib/spack/spack/directives.py:
```diff
@@ -784,6 +785,39 @@ def _execute_maintainer(pkg):
+        # and get discarded.
+        when_spec = make_when_spec(when)
+        if not when_spec:
+            return
```
(suggestion, no change required) Maybe slightly more clear:

```
if when is False:
    # Note that when is None, it means there is no condition (i.e. requirement applies unconditionally)
    return
when_spec = ...
```

alalazo replied:
Done in 222b6a7d6513fa00f1cdf807abdbae15cabf4314

scheibelp commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1263,36 +1316,53 @@ def provider_requirements(self):
-            rules = self._rules_from_requirements(virtual_str, requirements)
-            self.emit_facts_from_requirement_rules(rules, virtual=True)
+            rules = self._rules_from_requirements(
+                virtual_str, requirements, provenance=Provenance.PACKAGES_YAML_VIRTUAL
```
What's the purpose of a separate provenance for `PACKAGES_YAML_VIRTUAL` and `PACKAGES_YAML`?

Is it for more than identifying rules as having come from virtuals in `emit_facts_from_requirement_rules`? If not, I think this might be overloading the use of provenance and an extra parameter should be used instead.

alalazo replied:
> Is it for more than identifying rules as having come from virtuals in emit_facts_from_requirement_rules? 

No, it's for that. Here I package rules upfront in a data structure and process them afterwards. Why do you think the use is overloaded? What would you prefer instead? `VersionProvenance` vs. `RequirementProvenance`?

scheibelp replied:
> Why do you think the use is overloaded?

Provenance is intended as a component of multi-dimensional weighting (i.e. something that ends up as part of the facts in the final program). If you are just using it to change behavior outside of the program (when in fact the weights don't need to differ) then you are getting an unintended side effect as part of using this to influence other behavior. 

> What would you prefer instead? VersionProvenance vs. RequirementProvenance?

I would prefer not using the term "provenance", because idiomatically that is used in this code for weighting. I think it should be another property of a `RequirementRule`.

alalazo replied:
I can code another enum, it needs to distinguish between "package" labels (e.g. `hdf5`), the label "all" and "virtual" labels (e.g. `mpi`).

alalazo replied:
Renamed the attribute to "kind" in 7df01e92ad4f0c09be83301d0132309dbaf5dbdb and used a separate enum in 0d7087bcd936254974d1e985cfedd8ddb2b013b5

scheibelp commented on lib/spack/spack/solver/asp.py:
```diff
@@ -634,6 +643,12 @@ def raise_if_errors(self):
+#: Data class to collect information on a requirement
+RequirementRule = collections.namedtuple(
+    "RequirementRule", ["name", "policy", "requirements", "condition", "provenance", "message"]
+)
```
I'd prefer the `name` parameter be `pkg_name` to indicate it stores the name of a package (vs. someone might think it's a name for the rule).

alalazo replied:
Done in 72ba954451eb688447c9c6a7527e4fb68a30aa14

scheibelp commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -461,13 +461,58 @@ error(100, "Attempted to use external for '{0}' which does not satisfy any confi
+impose(Y) :- activate_requirement_rule(Y, Package, X).
+do_not_impose(Y) :-
+  not activate_requirement_rule(Y, Package, X),
+  requirement_group_member(Y, Package, X).
```
From experimenting this seems to be exclusively used to prompt the correct error message: I'm worried about complicating the `.lp` logic for that purpose without a holistic approach to handing error messages (the tests make sure we get this message, but is that truly the best message to get every time there is a failure and a requirement group is involved?).

alalazo replied:
> From experimenting this seems to be exclusively used to prompt the correct error message:

No, it's supposed to be part of the logic. A conditional requirement has the form (in the most general case with `one_of` or `any_of`):
```yaml
packages:
  zlib:
    require:
    - one_of: ["%gcc", "%clang", "%intel"]
      when: "@1.2.13"
```
The condition is the same for all the requirements, so if I look at the ASP facts for the yaml above I'll get:
```prolog
requirement_group("zlib",0).
requirement_policy("zlib",0,"one_of").
requirement_conditional("zlib",0).
condition(5,"None").
condition_requirement(5,"node","zlib").
condition_requirement(5,"node_version_satisfies","zlib","1.2.13").
imposed_constraint(5,"node_compiler_set","zlib","gcc").
imposed_constraint(5,"node_compiler_version_satisfies","zlib","gcc",":").
requirement_group_member(5,"zlib",0).
requirement_has_weight(5,0).
condition(6,"None").
condition_requirement(6,"node","zlib").
condition_requirement(6,"node_version_satisfies","zlib","1.2.13").
imposed_constraint(6,"node_compiler_set","zlib","clang").
imposed_constraint(6,"node_compiler_version_satisfies","zlib","clang",":").
requirement_group_member(6,"zlib",0).
requirement_has_weight(6,1).
condition(7,"None").
condition_requirement(7,"node","zlib").
condition_requirement(7,"node_version_satisfies","zlib","1.2.13").
imposed_constraint(7,"node_compiler_set","zlib","intel").
imposed_constraint(7,"node_compiler_version_satisfies","zlib","intel",":").
requirement_group_member(7,"zlib",0).
requirement_has_weight(7,2).
```
The `condition_requirement` is the same for each of the `one_of` requirements, so if one of them holds all of them hold[^1]. The logic here allows us to multiplex what we want to impose, and deactivate the rest.

[^1]: The `condition_requirement` maps to the single `when` in the object and defaults to "the node is present" for non-conditional requirements.

scheibelp replied:
Overall, I'm opposed to the notion of applying `condition_requirement` to all of these and then selectively deactivating all of them except one with `do_not_impose`: it would be more straightforward if the condition would "activate" the group.

alalazo replied:
I'll check how hard is to do that, but note that I am using the same mechanism that externals use to trim dependencies - so something we already do on ASP conditions.

scheibelp replied:
> I am using the same mechanism that externals use to trim dependencies

I acknowledge that `do_not_impose` is preexisting (EDIT), but I would like to avoid that if at all possible here. I personally would like to refactor it out entirely if possible (not that this is a request I'd make in this PR) but I'm more willing to believe that special treatment is required for externals given that they truncate dependencies.

alalazo replied:
Reworked using a single condition in edd5d89b3ba4028e5232a8f3a5bb525c06b53297 The facts for the case above now look like:
```prolog
requirement_group("zlib",0).
requirement_policy("zlib",0,"one_of").
condition(5,"condition to activate requirement 0").
condition_requirement(5,"node","zlib").
condition_requirement(5,"node_version_satisfies","zlib","0.2.13").
requirement_conditional("zlib",0,5).

condition(6,"None").
condition_requirement(6,"node","zlib").
condition_requirement(6,"node_compiler","zlib","gcc").
condition_requirement(6,"node_compiler_version_satisfies","zlib","gcc",":").
imposed_constraint(6,"node_compiler_set","zlib","gcc").
imposed_constraint(6,"node_compiler_version_satisfies","zlib","gcc",":").
requirement_group_member(6,"zlib",0).
requirement_has_weight(6,0).

condition(7,"None").
condition_requirement(7,"node","zlib").
condition_requirement(7,"node_compiler","zlib","clang").
condition_requirement(7,"node_compiler_version_satisfies","zlib","clang",":").
imposed_constraint(7,"node_compiler_set","zlib","clang").
imposed_constraint(7,"node_compiler_version_satisfies","zlib","clang",":").
requirement_group_member(7,"zlib",0).
requirement_has_weight(7,1).

condition(8,"None").
condition_requirement(8,"node","zlib").
condition_requirement(8,"node_compiler","zlib","intel").
condition_requirement(8,"node_compiler_version_satisfies","zlib","intel",":").
imposed_constraint(8,"node_compiler_set","zlib","intel").
imposed_constraint(8,"node_compiler_version_satisfies","zlib","intel",":").
requirement_group_member(8,"zlib",0).
requirement_has_weight(8,2).
```

scheibelp commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -461,13 +461,58 @@ error(100, "Attempted to use external for '{0}' which does not satisfy any confi
+      not requirement_conditional(Package, X),
+      requirement_group_member(Y, Package, X).
+
+% Avoid impose(Y) to be derived by other rules when it is part of a requirement
```
I'm not sure what this means and none of the `concretize_requirements.py` tests fail (including the new ones you added) when I remove this block.

alalazo replied:
This is an integrity constraint and expresses a property of the solution that shouldn't be violated. In this case, it should never happen that `impose(Y)` is in the answer set when `activate_requirement_rule(Y, Package, X)` is not.

alalazo replied:
Outdated because of edd5d89b3ba4028e5232a8f3a5bb525c06b53297

scheibelp commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -461,13 +461,58 @@ error(100, "Attempted to use external for '{0}' which does not satisfy any confi
+activate_requirement(Package, X) :-
+  attr("virtual_node", Package),
+  requirement_group(Package, X),
+  not requirement_conditional(Package, X).
```
It would be nice if we could derive a single fact from `node` or `virtual` node to reduce duplication:

```
attr("build", Package) :- attr("node", Package)
attr("build", Package) :- attr("virtual_node", Package)

activate_requirement(Package, X) :-
   attr("build", Package),
   requirement_group(Package, X),
   not requirement_conditional(Package, X).
```

alalazo replied:
Done in 5a9898d0b4fc8aa6c5265a06c8a56265327facd2

scheibelp commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -461,13 +461,58 @@ error(100, "Attempted to use external for '{0}' which does not satisfy any confi
+  attr("node", Package),
+  1 { condition_holds(Y) : requirement_group_member(Y, Package, X) },
+  requirement_group(Package, X),
+  requirement_conditional(Package, X).
```
(abstract design issue, more of a discussion prompt than a request) I find this confusing:

* My assumption is that `activate_requirement(Package, X)` means "the requirement group `X` for `Package` must be applied" (i.e. that one of the members must hold)
* This seems to say that in order for the requirement group X to be mandatory, that at least one of its members must hold

To me, the members of the requirement group are things that have to be applied only if we decide the requirement group is mandatory. I figured that would be achievable by replacing

```
activate_requirement_rules(Package) :- attr("node", Package).
```

with

```
activate_requirement_group(Package, X) :- 
  attr("node", Package),
  requirement_group(Package, X),
  requirement_condition(Z, X),
  condition_holds(Z).
```

the difference being that the `requirement_condition` Z is not a member of the `requirement_group` (well, it could be, but that would seem useless to me, like `require: +foo when: +foo`).

alalazo replied:
See https://github.com/spack/spack/pull/36286#discussion_r1165307712 For a conditional requirement all the conditions are equal and map to the `when=` argument. So saying:
```prolog
1 { condition_holds(Y) : requirement_group_member(Y, Package, X) },
```
means that the `when=` holds and we should consider the requirement.

alalazo replied:
Outdated because of edd5d89b3ba4028e5232a8f3a5bb525c06b53297

scheibelp reviewed (commented):
This is a preliminary review: there are many comments but the important ones (for now) are in `concretize.lp`. I plan on reading through this more in the next couple days but I think discussing these initial questions will help me with that.

alalazo committed:
Reflect in test name that we are testing conditional requirements

alalazo committed:
Add a unit test with two requirements in "one_of"

alalazo committed:
Fix wording in directives

alalazo committed:
Rename an attribute of RequirementRule

alalazo committed:
Distinguish among different kind of requirements using a new enum

alalazo committed:
Use a single condition to gate the requirement, instead of multiplexing

alalazo committed:
Reduce code duplication in logic program

alalazo committed:
Allow condition on variants that might not exists under "all"

alalazo committed:
Allow using "spec" with conditional requirements

This helps readability when there is a single constraint
to enforce conditionally.

alalazo committed:
Allow more complex requirements with directives

alalazo commented:
@scheibelp I also implemented the rewuest in https://github.com/spack/spack/pull/36286#issuecomment-1480590404 in https://github.com/spack/spack/pull/36286/commits/5bb5308e4b90fcf7fbd7034b7fd6a060d76cb560

alalazo commented:
Measured this PR again:

[radiuss-requires.csv](https://github.com/spack/spack/files/11323130/radiuss-requires.csv)
[radiuss-develop.csv](https://github.com/spack/spack/files/11323132/radiuss-develop.csv)
[radiuss.txt](https://github.com/spack/spack/files/11323133/radiuss.txt)

My laptop isn't "super-stable" (there are a few fluctuations, for good or bad) but it seems overall this PR has no performance effect, when requirements are not used:

![radiuss](https://user-images.githubusercontent.com/4199709/234297644-bbc4ecdc-d0c1-40cc-a8ae-554040a157bd.png)

scheibelp commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -2647,27 +2647,51 @@ Adding the following to a package:
 we express the fact that the current package *cannot be built* with the Intel
-compiler when we are trying to install a version "<=1.2". The ``when`` argument
-can be omitted, in which case the conflict will always be active.
-Conflicts are always evaluated after the concretization step has been performed,
```
Glad you caught that

scheibelp commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -494,12 +505,23 @@ requirement_weight(Package, Group, W) :-
   not requirement_group_satisfied(Package, X).
 
+
+error(10, Message) :-
```
(very minor) I think the weight of this should be close to the no-message one (definitely less, but making it this low feels like it's making it a whole different category of complaint)

alalazo replied:
I kept this ordering because of the following rationale:
1. Any valid ASP solution that results in an error is a failed concretization. Said otherwise, we can't have any `error` with any weight for a successful concretization. This means that any weight will preserve the correctness of concretization.
2. In the presence of an error, I'd like to prefer reporting requirements with custom messages, rather than other possible causes (e.g. having multiple compilers on a node), since it's likely they're more comprehensible to users. The hope is that users could diagnose a failed concretization better this way. To prefer reporting these errors, we need to assign them a lower weight.
3. Currently we have three weights for errors, which are `1`, `10` and `100`. I didn't want to introduce a new one in this PR.

scheibelp commented on lib/spack/spack/test/concretize_requirements.py:
```diff
@@ -586,3 +587,192 @@ def test_non_existing_variants_under_all(concretize_scope, mock_packages):
+):
+    """Test that a few properties of specs that are concretized with requirements
+    match the expectations.
+    """
```
Suggested reword: 

> Test that conditional requirements are required when the condition is met, and optional when the condition is not met

A couple things come to mind:

* IMO the most robust way to ensure that the requirement is in fact hard is to try to concretize a conflicting spec when the condition is met (i.e. force a concretizer error)
* As a sanity check, it *might* be worth checking that you can concretize something like `libelf@0.8.13%gcc` if you have `require: @0.8.13, when: %clang` (i.e. a requirement imposes a condition, we don't meet the condition, but still impose the constraint from the cmd line). This is less important than the behavior you checked, but if things go particularly wrong it may be worth testing.

alalazo replied:
Addressed in 57cd6ed

scheibelp commented on lib/spack/spack/test/concretize_requirements.py:
```diff
@@ -586,3 +587,192 @@ def test_non_existing_variants_under_all(concretize_scope, mock_packages):
+):
+    """Tests that we can express a requirement condition under 'all' also in cases where
+    the corresponding requirement doesn't exist. For those packages the requirement rule
+    is not emitted, since it can be determined to be always false.
```
I think when you say

> the corresponding requirement doesn't exist

you mean

> the corresponding condition spec mentions variants that don't exist in the package

If I got that right, I think you are referring to the dependencies of mpileaks (mpileaks itself appears to have the `+shared` variant, but packages it depends on do not); it might be worth explaining that in a comment.

It also might improve robustness to also assert:

```
assert not `~shared` in s['callpath']
```

That is more of a sanity check: if someone changes that package later, this test could suddenly not be asserting anything of interest.

alalazo replied:
Should be addressed in 57cd6ed

scheibelp commented on lib/spack/spack/solver/asp.py:
```diff
@@ -97,24 +99,38 @@ def getter(node):
 
-#: Look up version precedence strings by enum id
-version_origin_str = {i: name for i, name in enumerate(version_origin_fields)}
+class Provenance(enum.IntEnum):
```
I assume you specifically wanted to use `IntEnum` because these values are weights. I think that's worth explaining in a comment. I'm on the fence about using `enum.auto()` in this case: it obscures the importance of the ordering (readers should be made aware that the order these attributes are declared in the code is essential).

alalazo replied:
> I assume you specifically wanted to use IntEnum because these values are weights. 

They are not weights directly, but they are used to determine weights. Specifically, in:

https://github.com/spack/spack/blob/5bb5308e4b90fcf7fbd7034b7fd6a060d76cb560/lib/spack/spack/solver/asp.py#L906-L913

we sort versions by provenance first. A "regular" enum doesn't define the `__lt__` operator that is needed for sorting, while an int enum does it for free.

scheibelp commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1266,26 +1333,54 @@ def provider_requirements(self):
+                continue
+
+            # Write explicitly if a requirement is conditional or not
+            if rule.condition and main_requirement_condition != spack.spec.Spec():
```
I'm trying to figure out why the `rule.condition` check is needed: is that if `condition is True` (i.e. there is no condition, always apply the group)?

I think it might be slightly clearer to make this explicit with something like

```
always_off = False
sometimes_on = False
always_on = False

if rule.condition is False or main_requirement_condition is False:
    always_off = True
if rule.condition is True or main_requirement_condition != spack.spec.Spec():
    always_on = True
else:
    sometimes_on = True
```

if that doesn't appeal, I'll suggest that it would be nice to clarify this *somehow*.

alalazo replied:
I think it's redundant in the current formulation, so I'll drop that.[^1]

[^1]: it was to ensure we treated the `rule.condition = None` case the same as `rule.condition = True`, but this is already accounted for in `make_when_spec`.

alalazo replied:
Dropped in 57cd6ed

scheibelp commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1011,35 +1030,83 @@ def package_compiler_defaults(self, pkg):
+                        pkg_name=pkg.name,
+                        policy=policy,
+                        requirements=requirements,
+                        kind=Provenance.PACKAGE_PY,
```
(bug) This should be a `RequirementKind`

alalazo replied:
Fixed in 3eeb585

scheibelp reviewed (changes requested):
Overall this looks great. All of my comments are related to clarity except one which I prefixed with "(bug)".

alalazo committed:
Fix bug (wrong type for RequirementRule.kind)

alalazo committed:
Address review comments
